source/lac/trilinos_vector.cc

```diff
@@ -562,24 +578,11 @@ namespace TrilinosWrappers
 
   void
   Vector::reinit (const size_type n,
-                  const bool      omit_zeroing_entries)
+                  const bool      )
   {
-    if (size() != n)
-      {
-        Epetra_LocalMap map ((TrilinosWrappers::types::int_type)n, 0,
-                             Utilities::Trilinos::comm_self());
-        vector.reset (new Epetra_FEVector (map));
-      }
-    else if (omit_zeroing_entries == false)
-      {
-        int ierr;
-        ierr = vector->GlobalAssemble(last_action);
-        (void)ierr;
-        Assert (ierr == 0, ExcTrilinosError(ierr));
-
-        ierr = vector->PutScalar(0.0);
-        Assert (ierr == 0, ExcTrilinosError(ierr));
-      }
+    Epetra_LocalMap map ((TrilinosWrappers::types::int_type)n, 0,
+                         Utilities::Trilinos::comm_self());
+    vector.reset (new Epetra_FEVector (map));
```

Comment:
There was, as I observed when running through the debugger. The problem is GlobalAssemble goes through the communicator. Even though we deal with a local map, the communicator is still MPI_Comm and could do strange things. And even if we did omit zeroing entries, we could end up with a vector that we thought was initialized correctly but later we do something with the communicator that goes wrong.

Moreover, I'm not sure whether we need to provide a fast path when initializing from a size (or Epetra_Map / IndexSet in the MPI case) - I effectively disable that part. (We can change to a similar structure as above, but I decided that the additional code is probably not worth it.)